File: src/dataset/typescript/complex-numbers_exercism-wackerow/openAI/ChatGPT4_complex-numbers_v4.ts
export default class ComplexNumber {
  private readonly _abs: number;
  private readonly _conj: ComplexNumber;
  private readonly _exp: ComplexNumber;

  constructor(private readonly _real: number, private readonly _imag: number) {
    this._abs = Math.hypot(_real, _imag);
    this._conj = new ComplexNumber(_real, -_imag);
    const expReal = Math.exp(_real);
    this._exp = new ComplexNumber(expReal * Math.cos(_imag), expReal * Math.sin(_imag));
  }

  get real(): number {
    return this._real;
  }

  get imag(): number {
    return this._imag;
  }

  get abs(): number {
    return this._abs;
  }

  get conj(): ComplexNumber {
    return this._conj;
  }

  get exp(): ComplexNumber {
    return this._exp;
  }

  public add(that: ComplexNumber): ComplexNumber {
    return new ComplexNumber(this._real + that._real, this._imag + that._imag);
  }

  public sub(that: ComplexNumber): ComplexNumber {
    return new ComplexNumber(this._real - that._real, this._imag - that._imag);
  }

  public mul(that: ComplexNumber): ComplexNumber {
    const real = this._real * that._real - this._imag * that._imag;
    const imag = this._imag * that._real + this._real * that._imag;
    return new ComplexNumber(real, imag);
  }

  public div(that: ComplexNumber): ComplexNumber {
    const denom = that._real * that._real + that._imag * that._imag;
    const real = (this._real * that._real + this._imag * that._imag) / denom;
    const imag = (this._imag * that._real - this._real * that._imag) / denom;
    return new ComplexNumber(real, imag);
  }
}

File: src/dataset/python/complex-numbers_exercism-fruit-in/openAI/ChatGPT4_complex_numbers_v4.py
import math


class ComplexNumber:
    __slots__ = ('real', 'imaginary')

    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        return ComplexNumber(self.real + other.real, self.imaginary + other.imaginary)

    def __mul__(self, other):
        return ComplexNumber(
            self.real * other.real - self.imaginary * other.imaginary,
            self.imaginary * other.real + self.real * other.imaginary
        )

    def __sub__(self, other):
        return ComplexNumber(self.real - other.real, self.imaginary - other.imaginary)

    def __truediv__(self, other):
        denom = other.real * other.real + other.imaginary * other.imaginary
        return ComplexNumber(
            (self.real * other.real + self.imaginary * other.imaginary) / denom,
            (self.imaginary * other.real - self.real * other.imaginary) / denom
        )

    def __abs__(self):
        return math.hypot(self.real, self.imaginary)

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        exp_real = math.exp(self.real)
        return ComplexNumber(
            exp_real * math.cos(self.imaginary),
            exp_real * math.sin(self.imaginary)
        )